temporal-sdk/src/main/java/io/temporal/internal/sync/POJOWorkflowImplementationFactory.java

```diff
@@ -94,23 +97,33 @@
     this.contextPropagators = contextPropagators;
   }
 
-  void addWorkflowImplementationTypes(
+  void registerWorkflowImplementationTypes(
       WorkflowImplementationOptions options, Class<?>[] workflowImplementationTypes) {
     for (Class<?> type : workflowImplementationTypes) {
-      addWorkflowImplementationType(options, type);
+      registerWorkflowImplementationType(options, type);
     }
   }
 
   <R> void addWorkflowImplementationFactory(Class<R> clazz, Functions.Func<R> factory) {
+    @SuppressWarnings("unchecked")
     WorkflowImplementationOptions unitTestingOptions =
         WorkflowImplementationOptions.newBuilder()
             .setFailWorkflowExceptionTypes(Throwable.class)
             .build();
     addWorkflowImplementationFactory(unitTestingOptions, clazz, factory);
   }
 
+  @SuppressWarnings("unchecked")
   <R> void addWorkflowImplementationFactory(
       WorkflowImplementationOptions options, Class<R> clazz, Functions.Func<R> factory) {
+    if (DynamicWorkflow.class.isAssignableFrom(clazz)) {
+      if (dynamicWorkflowImplementationFactory != null) {
+        throw new IllegalStateException(
+            "An implementation of DynamicWorkflow or its factory is already registered with the worker");
```

Comment:
Same as other comment - would be nice to expose this restriction to user in docstring.